e2e/LongHaul/service/Program.cs

```diff
@@ -0,0 +1,117 @@
+            Parameters parameters = null;
+            ParserResult<Parameters> result = Parser.Default.ParseArguments<Parameters>(args)
+                .WithParsed(parsedParams =>
+                {
+                    parameters = parsedParams;
+                })
+                .WithNotParsed(errors =>
+                {
+                    Environment.Exit(1);
+                });
+
+            s_logger = InitializeLogging(parameters);
+
+            // Log system health before initializing hub
+            SystemHealthMonitor.BuildAndLogSystemHealth(s_logger);
+
+            s_logger.Event(StartingRun);
+
+            var iotHub = new IotHub(
+                s_logger,
+                parameters.IoTHubConnectionString,
+                parameters.DeviceId,
+                parameters.TransportProtocol);
+
+            // Log system health after initializing hub
+            SystemHealthMonitor.BuildAndLogSystemHealth(s_logger);
+            iotHub.Initialize();
+
+            // Log system health after opening connection to hub
+            SystemHealthMonitor.BuildAndLogSystemHealth(s_logger);
+
+            using CancellationTokenSource cancellationTokenSource = ConfigureAppExit();
+
+            try
+            {
+                await iotHub.RunAsync(cancellationTokenSource.Token).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException) { } // user signalled an exit
+            catch (Exception ex)
+            {
+                s_logger.Trace($"Device app failed with exception {ex}");
+            }
+
+            iotHub.Dispose();
+
+            // Log system health after disposing hub
+            SystemHealthMonitor.BuildAndLogSystemHealth(s_logger);
+
+            s_logger.Flush();
+            s_aiLoggingProvider.Dispose();
+        }
+
+        private static Logger InitializeLogging(Parameters parameters)
+        {
+            var helper = new IotHubConnectionStringHelper(parameters.IoTHubConnectionString);
+            var logBuilder = new LoggingBuilder
+            {
+                AppContext =
+                {
+                    { Hub, helper.HostName },
+                    { Transport, $"HTTP/{parameters.TransportProtocol}" },
+                },
+            };
+            foreach (KeyValuePair<string, string> kvp in s_commonProperties)
+            {
+                logBuilder.AppContext.Add(kvp.Key, kvp.Value);
+            }
+            logBuilder.LogProviders.Add(new ConsoleLogProvider { ShouldLogContext = false, ShouldUseColor = true });
+            s_aiLoggingProvider = new ApplicationInsightsLoggingProvider(parameters.InstrumentationKey);
+            logBuilder.LogProviders.Add(s_aiLoggingProvider);
+
+            Logger logger = logBuilder.BuildLogger();
+            return logger;
+        }
+
+        private static CancellationTokenSource ConfigureAppExit()
+        {
+            var cancellationTokenSource = new CancellationTokenSource();
```

Comment:
This will be leaked as we aren't disposing it (nor can we if it is local to this method).

Maybe set it on the class so the main method can dispose it?

We'll want to do some diagnostics to detect memory leaks on exit.